pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -225,27 +230,50 @@ func configFromEnv() (cfg Config, ok bool) {
 		cfg.Global.DomainName = os.Getenv("OS_USER_DOMAIN_NAME")
 	}
 
-	ok = cfg.Global.AuthURL != "" &&
-		cfg.Global.Username != "" &&
-		cfg.Global.Password != "" &&
-		(cfg.Global.TenantID != "" || cfg.Global.TenantName != "" ||
-			cfg.Global.DomainID != "" || cfg.Global.DomainName != "" ||
-			cfg.Global.Region != "" || cfg.Global.UserID != "" ||
-			cfg.Global.TrustID != "")
-
 	cfg.Metadata.SearchOrder = fmt.Sprintf("%s,%s", configDriveID, metadataID)
 	cfg.Metadata.DHCPDomain = "novalocal"
 	cfg.BlockStorage.BSVersion = "auto"
-
-	return
 }
 
-func readConfig(config io.Reader) (Config, error) {
-	if config == nil {
-		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
+func configFromSecret(cfg *Config, kubeClient clientset.Interface, name string, namespace string) error {
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		for name, data := range secret.Data {
```

Comment:
I don't know enough about these options to know if it's reasonable to mix these with the global config... if this override is provided, should we start by emptying out these values from the global config? The scenarios I'm thinking of most are:
* mutually-exclusive options (e.g. global config specifies tenant-id, local config specifies tenant-name). I assume there's a precedence order the library uses those in, and it seems weird to require the local config to null out the global value to ensure it's value takes effect
* reuse of globally provided credentials. I can't tell if this is a feature or a security hole, but as is, this would let you use globally configured credentials against a different tenant/domain/region, right?

Reply:
An the other point is, the global config has the admin credentials, we are really trying to help end users use their own credentials (and their own tenant). So in effect we are trying to get them to limit themselves ...

Reply:
> An the other point is, the global config has the admin credentials, we are really trying to help end users use their own credentials (and their own tenant). So in effect we are trying to get them to limit themselves ...

it has the global credentials, but only pointed at the tenant and region the cluster admin selected. I'm not sure we can assume a random pod author should be able to make use of those credentials to choose a different tenant or region, right?

Reply:
Good point, if they specify a tenant or region, then i will make sure they should specify a different userid/name (and hence the associated domain and password)